### ISSUE TITLE
Find prefixes from prefix.cc in DCTAP

### ISSUE THREAD

**Issue by labra** [rudof-project/rudof#97]
At this moment DCTAP conversion is based on a fixed set of prefix declarations which can be injected through a config file. 

At suggested by @jonassmedegaard the usability could be improved by retrieving the prefix declarations from http://prefix.cc or by using a cached version of those prefix declarations.

**Comment by kcoyle** [comment]
My conversion of a DCTAP failed on prefixes, so I decided to pre-expand the prefixes before running rudof. Then I got this:
```
Error: PrefixMap error: Prefix 'http' not found in PrefixMap 'prefix dc: <http://purl.org/dc/elements/1.1/>
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix : <http://example.org/>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix xsd: <http://www.w3.org/2001/XMLSchema#>
' at line: 3, field: http://purl.org/dc/terms/type
```
It is going to be very hard to have all of the prefixes that people might use in a DCTAP. We have recognized that there needs to be a way to have a file of the prefixes used, but we haven't yet figured that out. Meanwhile, adding http and https to the prefixMap would allow us to experiment with rudof.

**Comment by jonassmedegaard** [comment]
To me this looks like a parsing bug in Rudof:
The failing RDF statement at line 3 - `prefix : <http://example.org/>` does *not* have prefix prefix, it has *no* prefix.

Therefore your comment, @kcoyle, seems independent from the error message above your comment - if I understand you correctly.

**Comment by kcoyle** [comment]
I assumed that the reference to line 3 was line 3 of the dctap. The first three lines of that are:

shapeID,propertyLabel,propertyID,valueNodeType,valueDatatype,valueShape,valueConstraint,valueConstraintType,note
,,,,,,,,
ScholarlyResource,Type,http://purl.org/dc/terms/type,IRI,,,http://purl.org/coar/resource_type/,IRIstem,"Resource type, using the COAR Resorce Types vocabulary."

The config file cannot have http://purl.org/dc/terms/type in it. "/type" is an actual property, beyond the prefix, so that wouldn't be in the prefix file. It also seems that the set of prefixes being used is what was echo'd on the error message. I note that there is a larger prefix file (probably using prefixCC) [here](https://github.com/rudof-project/rudof/blob/master/dctap/config/prefix_cc_context.jsonld) but it doesn't seem to be implemented yet. In any case, it does seem that one should allow full IRIs with http or https.

**Comment by jonassmedegaard** [comment]
Ah, with that additional background information - and noticing the last line of the error message - I agree with your interpretation of it.  Sorry for my confusion before.

I don't follow your remark that `http://purl.org/dc/terms/type` cannot be the config file - a lookup at https://prefix.cc indicates that it is used with the CURIE [dc:type](https://prefix.cc/dc:type).

Yes, I agree with you that full IRIs are legal values for PropertyID, as described [in the DCTAP primer](https://www.dublincore.org/specifications/dctap/primer/#property-identifier)

**Comment by labra** [comment]
> My conversion of a DCTAP failed on prefixes, so I decided to pre-expand the prefixes before running rudof. Then I got this:
> 
> ```
> Error: PrefixMap error: Prefix 'http' not found in PrefixMap 'prefix dc: <http://purl.org/dc/elements/1.1/>
> prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
> prefix : <http://example.org/>
> prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
> prefix xsd: <http://www.w3.org/2001/XMLSchema#>
> ' at line: 3, field: http://purl.org/dc/terms/type
> ```

Indeed, that looks like a bug in the parser that thinks that `http:`  is the start of a curie like `xsd:string`. Could you send me a simplified snippet of the DCTAP file that you are using so I can reproduce the error?

> It is going to be very hard to have all of the prefixes that people might use in a DCTAP. We have recognized that there needs to be a way to have a file of the prefixes used, but we haven't yet figured that out. Meanwhile, adding http and https to the prefixMap would allow us to experiment with rudof.

Indeed, currently `rudof` takes as input a config file which can contain a list of prefix declarations. That config file can be changed and customized by the users. 

Another possibility is that `rudof` uses also the default prefix declarations which can be downloaded from [prefix.cc](https://prefix.cc/). We started to implement this option, but it is not finished yet. 
I keep this issue open as a reminded of that possibility, but I will definitely open a new issue if I can reproduce the error with the `http:` prefix.
